src/transformers/models/zamba/modeling_zamba.py

```diff
@@ -0,0 +1,1728 @@
+    and is why Mamba is called **selective** state spaces)
+
+    This module differs from `transformers.models.mamba.modeling_mamba.MambaMixer` in two ways:
+    - Added multi-head: the output of `self.in_proj` is split into `self.n_mamba_heads` heads, and each head
+    undergoes an independent forward pass, identical to the original `MambaMixer`, up until the pre-activations of
+    `self.out_proj`. The pre-activations, coming from different mamba heads, are then concatenated and fed into `self.out_proj`.
+    - Added `attention_mask` for batched inference: this tensor multiplies input and output of the convolution layer, setting
+    to zero embeddings associated with `attention_mask == 0` thus preventing the layer to attend to such tokens.
+    """
+
+    def __init__(self, config: ZambaConfig, layer_idx):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.hidden_size = config.hidden_size
+        self.ssm_state_size = config.mamba_d_state
+        self.conv_kernel_size = config.mamba_d_conv
+        self.intermediate_size = config.mamba_expand * config.hidden_size
+        self.time_step_rank = config.mamba_dt_rank
+        self.n_mamba_heads = config.n_mamba_heads
+        self.use_conv_bias = config.mamba_conv_bias
+        self.use_bias = config.mamba_proj_bias
+        self.conv1d = nn.Conv1d(
+            in_channels=self.intermediate_size,
+            out_channels=self.intermediate_size,
+            bias=self.use_conv_bias,
+            kernel_size=self.conv_kernel_size,
+            groups=self.intermediate_size,
+            padding=self.conv_kernel_size - 1,
+        )
+
+        self.activation = config.hidden_mamba_act
+        self.act = ACT2FN[config.hidden_mamba_act]
+
+        self.use_fast_kernels = config.use_mamba_kernels
+
+        assert self.intermediate_size % self.n_mamba_heads == 0, '`intermediate_size` should be divisible by `n_mamba_heads`.'
+
+        # projection of the input hidden states
+        self.in_proj = nn.Linear(self.hidden_size, self.intermediate_size * 2, bias=self.use_bias)
+        # selective projection used to make dt, B and C input dependent
+        self.x_proj_weight = nn.Parameter(
+            (
+                torch.zeros(
+                    self.n_mamba_heads,
+                    self.time_step_rank + self.ssm_state_size * 2,
+                    self.intermediate_size // self.n_mamba_heads,
+                )
+            )
+        )
+        # time step projection (discretization)
+        self.dt_proj_weight = nn.Parameter(
+            (torch.zeros(self.n_mamba_heads, self.intermediate_size // self.n_mamba_heads, self.time_step_rank) - 0.5)
+            * 2
+            / self.time_step_rank**0.5
+        )  # (h d dt_rank)
+        self.dt_proj_bias = nn.Parameter(
+            torch.zeros(self.n_mamba_heads, self.intermediate_size // self.n_mamba_heads)
+        )  # (h d)
```

Comment:
let's improve readability of these please! comment on top, define shapes like self.intermediate_size // self.n_mamba_heads that are re-used

Reply:
defined `self.mamba_head_dim` here https://github.com/Zyphra/transformers_zamba/blob/97c646c2068a2b254c53518698c62ebc8fea5645/src/transformers/models/zamba/modeling_zamba.py#L577 and updated all occurrences of this quantity. We also slightly updated the comments on top of this block https://github.com/Zyphra/transformers_zamba/blob/97c646c2068a2b254c53518698c62ebc8fea5645/src/transformers/models/zamba/modeling_zamba.py#L596. Please let me know if more needs to be done here